src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt

```diff
@@ -0,0 +1,174 @@
+import io.smallrye.mutiny.Uni
+import io.smallrye.mutiny.coroutines.asUni
+import io.vertx.core.Vertx
+import io.vertx.kotlin.coroutines.dispatcher
+import jakarta.annotation.security.RolesAllowed
+import jakarta.enterprise.context.RequestScoped
+import jakarta.inject.Inject
+import jakarta.ws.rs.core.Response
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+import java.util.*
+
+/**
+ * Task comments API implementation
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+@RequestScoped
+@WithSession
+class TaskCommentsApiImpl : TaskCommentsApi, AbstractApi() {
+
+    @Inject
+    lateinit var taskCommentController: TaskCommentController
+
+    @Inject
+    lateinit var taskController: TaskController
+
+    @Inject
+    lateinit var taskCommentTranslator: TaskCommentTranslator
+
+    @Inject
+    lateinit var vertx: Vertx
+
+    @RolesAllowed(UserRole.ADMIN.NAME, UserRole.USER.NAME)
+    override fun listTaskComments(
+        projectId: UUID,
+        milestoneId: UUID,
+        taskId: UUID,
+        first: Int?,
+        max: Int?
+    ): Uni<Response> = CoroutineScope(vertx.dispatcher()).async {
+        val userId = loggedUserId ?: return@async createUnauthorized(UNAUTHORIZED)
+
+        val (projectMilestone, errorResponse) = getProjectMilestoneAccessRights(projectId, milestoneId, userId)
+        if (errorResponse != null) return@async errorResponse
+        val task = taskController.find(projectMilestone!!.first, taskId) ?: return@async createNotFound(
+            createNotFoundMessage(TASK, taskId)
+        )
+
+        val (comments, count) = taskCommentController.listTaskComments(task, first, max)
+        createOk(taskCommentTranslator.translate(comments), count)
+    }.asUni()
+
+    @WithTransaction
+    @RolesAllowed(UserRole.ADMIN.NAME, UserRole.USER.NAME)
+    override fun createTaskComment(
+        projectId: UUID,
+        milestoneId: UUID,
+        taskId: UUID,
+        taskComment: TaskComment
+    ): Uni<Response> = CoroutineScope(vertx.dispatcher()).async {
+        val userId = loggedUserId ?: return@async createUnauthorized(UNAUTHORIZED)
+        if (taskComment.taskId != taskId) return@async createBadRequest("Task id in path and body do not match")
+        val (projectMilestone, errorResponse) = getProjectMilestoneAccessRights(projectId, milestoneId, userId)
+        if (errorResponse != null) return@async errorResponse
+
+        val task = taskController.find(projectMilestone!!.first, taskId) ?: return@async createNotFound(
+            createNotFoundMessage(TASK, taskId)
+        )
+
+        taskComment.referencedUsers.forEach { referencedUserId ->
+            if (!projectController.hasAccessToProject(task.milestone.project, referencedUserId)) {
+                return@async createBadRequest("Referenced user $referencedUserId is not an assignee of the task")
+            }
+        }
```

Comment:
So we can not reference users that are not assigned to this task?

Reply:
No i just noticed it's wrong comment. We can't reference user that has no rights for the project

Reply:
Fixed the message